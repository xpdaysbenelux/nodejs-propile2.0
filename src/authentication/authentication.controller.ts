import {
    Controller,
    Post,
    Body,
    UseGuards,
    HttpCode,
    HttpStatus,
    Get,
    Req,
    Res,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { Response } from 'express';

import {
    ResetPasswordRequest,
    LoginRequest,
    AuthenticationUserResponse,
    RequestPasswordResetRequest,
} from './dto';
import { AuthenticationService } from './authentication.service';
import { IUserSession } from '../_shared/constants';
import { AuthenticationQueries } from './authentication.queries';
import { LoginGuard, AuthenticatedGuard } from '../_shared/guards';
import { UserSession, Origin } from '../_shared/decorators';
import { destroyExpressSession } from '../session.middleware';

@ApiTags('auth')
@Controller('auth')
export class AuthenticationController {
    constructor(
        private readonly authService: AuthenticationService,
        private readonly authQueries: AuthenticationQueries,
    ) {}

    @UseGuards(LoginGuard)
    @HttpCode(HttpStatus.OK)
    @Post('login')
    login(
        // body is not used, but here for swagger docs
        @Body() _body: LoginRequest,
        @UserSession() session: IUserSession,
    ): Promise<AuthenticationUserResponse> {
        return this.authQueries.getAuthenticatedUser(session.userId);
    }

    @UseGuards(AuthenticatedGuard)
    @Get('authenticate')
    authenticate(
        @UserSession() session: IUserSession,
    ): Promise<AuthenticationUserResponse> {
        return this.authQueries.getAuthenticatedUser(session.userId);
    }

    @UseGuards(AuthenticatedGuard)
    @HttpCode(HttpStatus.NO_CONTENT)
    @Post('logout')
    async logout(
        // Using any because the build fails when using passport types for the request
        // eslint-disable-next-line
        @Req() request: any,
        @Res() response: Response,
    ): Promise<void> {
        await destroyExpressSession(request, response);
        response.send();
    }

    @HttpCode(HttpStatus.NO_CONTENT)
    @Post('request-password-reset')
    requestPasswordReset(
        @Body() body: RequestPasswordResetRequest,
        @Origin() origin: string,
    ): Promise<void> {
        return this.authService.requestPasswordReset(body, origin);
    }

    @HttpCode(HttpStatus.NO_CONTENT)
    @Post('reset-password')
    resetPassword(@Body() body: ResetPasswordRequest): Promise<void> {
        return this.authService.resetPassword(body);
    }
}
