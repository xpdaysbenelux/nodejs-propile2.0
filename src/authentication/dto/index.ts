export * from './authenticated-user.dto';
export * from './login.dto';
export * from './password.dto';
export * from './request-password-reset.dto';
export * from './reset-password.dto';
