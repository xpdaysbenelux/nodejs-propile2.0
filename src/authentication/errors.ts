import { BadRequestException } from '@nestjs/common';

export class ResetTokenInvalid extends BadRequestException {
    constructor() {
        super('The given resetToken is not valid', 'RESET_TOKEN_INVALID');
    }
}

export class ResetTokenExpired extends BadRequestException {
    constructor() {
        super('The given resetToken is expired', 'RESET_TOKEN_EXPIRED');
    }
}
