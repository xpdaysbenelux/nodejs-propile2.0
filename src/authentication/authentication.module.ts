import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';

import { AuthenticationController } from './authentication.controller';
import { AuthenticationService } from './authentication.service';
import { PassportLocalStrategy } from './passport-local.strategy';
import { DatabaseModule } from '../database';
import { Config } from '../config';
import { UsersModule } from '../users/users.module';
import { AuthenticationQueries } from './authentication.queries';
import { SessionSerializer } from './session.serializer';
import { MailerModule } from '../mailer/mailer.module';

@Module({
    imports: [
        DatabaseModule,
        JwtModule.register({
            secret: Config.jwt.secret,
        }),
        UsersModule,
        MailerModule,
    ],
    controllers: [AuthenticationController],
    providers: [
        AuthenticationService,
        AuthenticationQueries,
        PassportLocalStrategy,
        SessionSerializer,
    ],
})
export class AuthenticationModule {}
