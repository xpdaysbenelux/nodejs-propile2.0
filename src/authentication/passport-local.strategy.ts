import { Strategy } from 'passport-local';
import { PassportStrategy } from '@nestjs/passport';
import { Injectable } from '@nestjs/common';

import { AuthenticationService } from './authentication.service';
import { IUserSession } from '../_shared/constants';
import { AuthenticationQueries } from './authentication.queries';

@Injectable()
export class PassportLocalStrategy extends PassportStrategy(Strategy) {
    constructor(
        private readonly authService: AuthenticationService,
        private readonly authQueries: AuthenticationQueries,
    ) {
        super();
    }

    async validate(username: string, password: string): Promise<IUserSession> {
        const user = await this.authService.login(username, password);
        return this.authQueries.composeUserSession(user.id);
    }
}
