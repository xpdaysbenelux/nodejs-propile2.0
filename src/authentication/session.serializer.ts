import { PassportSerializer } from '@nestjs/passport';
import { Injectable } from '@nestjs/common';

import { IUserSession } from '../_shared/constants';
import { AuthenticationQueries } from './authentication.queries';

@Injectable()
export class SessionSerializer extends PassportSerializer {
    constructor(private readonly authQueries: AuthenticationQueries) {
        super();
    }

    serializeUser(
        session: IUserSession,
        done: (err: Error, session: string) => void,
    ): void {
        done(null, session.userId);
    }

    async deserializeUser(
        userId: string,
        done: (err: Error, user?: IUserSession) => void,
    ): Promise<void> {
        try {
            const session = await this.authQueries.composeUserSession(userId);
            done(null, session);
        } catch (error) {
            done(error);
        }
    }
}
