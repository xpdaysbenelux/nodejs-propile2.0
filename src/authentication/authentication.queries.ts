import { Injectable } from '@nestjs/common';

import { UserRepository } from '../database';
import { IUserSession } from '../_shared/constants';
import { AuthenticationUserResponse } from './dto';
import {
    permissionsFromRoles,
    createDefaultPermissions,
} from '../_util/permissions.helper';

@Injectable()
export class AuthenticationQueries {
    constructor(private readonly userRepository: UserRepository) {}

    async composeUserSession(userId: string): Promise<IUserSession> {
        const user = await this.userRepository.findOne(userId, {
            relations: ['roles'],
        });
        if (!user) return null;
        return {
            userId: user.id,
            email: user.email,
            state: user.state,
            firstName: user.firstName,
            lastName: user.lastName,
            permissions: createDefaultPermissions(
                permissionsFromRoles(user.roles),
            ),
        };
    }

    async getAuthenticatedUser(
        userId: string,
    ): Promise<AuthenticationUserResponse> {
        const user = await this.userRepository
            .createQueryBuilder('user')
            .select([
                'user.id',
                'user.createdAt',
                'user.updatedAt',
                'user.createdBy',
                'user.updatedBy',
                'user.email',
                'user.state',
                'user.firstName',
                'user.lastName',
                'role.permissions',
            ])
            .innerJoin('user.roles', 'role')
            .where('user.id = :userId', { userId })
            .getOne();
        if (!user) return undefined;
        const { roles, ...otherProps } = user;
        return {
            ...otherProps,
            permissions: permissionsFromRoles(roles),
        };
    }
}
