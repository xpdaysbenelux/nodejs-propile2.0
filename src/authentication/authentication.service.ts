import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { TokenExpiredError } from 'jsonwebtoken';
import * as bcrypt from 'bcrypt';

import { UserRepository, User } from '../database';
import { ResetPasswordRequest, RequestPasswordResetRequest } from './dto';
import { ResetTokenInvalid, ResetTokenExpired } from './errors';
import { UserState } from '../_shared/constants';
import { MailerService } from '../mailer/mailer.service';
import { requestPasswordResetMessage } from '../mailer/messages';
import { canActivateWithUserState } from '../_shared/guards';

@Injectable()
export class AuthenticationService {
    constructor(
        private readonly userRepository: UserRepository,
        private readonly jwtService: JwtService,
        private readonly mailerService: MailerService,
    ) {}

    async login(email: string, password: string): Promise<User> {
        // Try to find the user
        const user = await this.userRepository.findOne({ email });
        if (!user) {
            throw new UnauthorizedException();
        }

        // Check if user is active
        canActivateWithUserState(user.state, [UserState.Active]);

        // Given password should match the stored hashed password
        const match = await bcrypt.compare(password, user.password);
        if (!match) {
            throw new UnauthorizedException();
        }

        return user;
    }

    async requestPasswordReset(
        body: RequestPasswordResetRequest,
        origin: string,
    ): Promise<void> {
        const { email } = body;

        // If the user is not found, just do nothing
        const user = await this.userRepository.findOne({ email });
        if (!user) {
            console.info(
                `Request password reset done for unknown email: ${email}`,
            );
            return;
        }

        // Add resetToken to the user
        const resetToken = await this.jwtService.signAsync(
            { email },
            { expiresIn: '1d' },
        );
        user.resetToken = resetToken;
        user.updatedBy = user.email;
        await this.userRepository.save(user);

        // Send mail to inform user
        this.mailerService.sendMail(
            requestPasswordResetMessage(email, resetToken, origin),
        );
    }

    async resetPassword(body: ResetPasswordRequest): Promise<void> {
        const { newPassword, resetToken } = body;

        // Check if token is still valid
        try {
            await this.jwtService.verifyAsync(resetToken);
        } catch (error) {
            console.error(error);
            if (error instanceof TokenExpiredError) {
                throw new ResetTokenExpired();
            }
            throw new ResetTokenInvalid();
        }

        // Check if token is linked to a user
        const decoded = this.jwtService.decode(resetToken) as { email: string };
        const user = await this.userRepository.findOne({
            resetToken,
        });
        if (!user || user.email !== decoded.email) {
            throw new ResetTokenInvalid();
        }

        // Check if the user is not inactive
        canActivateWithUserState(user.state, [
            UserState.Active,
            UserState.Registering,
        ]);

        // Update the user in the database
        const hashedPassword = await this.hashPassword(newPassword);
        user.state = UserState.Active;
        user.password = hashedPassword;
        user.resetToken = null;
        user.updatedBy = user.email;

        await this.userRepository.save(user);
    }

    hashPassword(password: string): Promise<string> {
        return bcrypt.hash(password, 10);
    }
}
