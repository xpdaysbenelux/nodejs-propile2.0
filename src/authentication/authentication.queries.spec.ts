import { Test, TestingModule } from '@nestjs/testing';

import { AuthenticationQueries } from './authentication.queries';
import { DatabaseModule, UserRepository } from '../database';

describe('AuthenticationQueries', () => {
    let userRepository: UserRepository;
    let authQueries: AuthenticationQueries;

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            imports: [DatabaseModule],
            providers: [AuthenticationQueries],
        }).compile();

        authQueries = module.get(AuthenticationQueries);
        userRepository = module.get(UserRepository);
    });

    afterAll(() => {
        userRepository.manager.connection.close();
    });

    describe('composeUserSession', () => {
        it(`should return the correct user`, async () => {
            const userId = 'c4cb4582-1e97-4e3e-9d49-c744c8c1a229';
            const result = await authQueries.composeUserSession(userId);
            expect(result).toMatchSnapshot();
        });

        it('should return no user if not found', async () => {
            const userId = '63440b0f-59b3-473e-bf87-2c05136cc005';
            const result = await authQueries.composeUserSession(userId);
            expect(result).toMatchSnapshot();
        });
    });

    describe('getLoggedInUser', () => {
        it(`should return the correct user`, async () => {
            const userId = 'c4cb4582-1e97-4e3e-9d49-c744c8c1a229';
            const result = await authQueries.getAuthenticatedUser(userId);
            expect(result).toMatchSnapshot();
        });

        it('should return no user if not found', async () => {
            const userId = '63440b0f-59b3-473e-bf87-2c05136cc005';
            const result = await authQueries.getAuthenticatedUser(userId);
            expect(result).toMatchSnapshot();
        });
    });
});
