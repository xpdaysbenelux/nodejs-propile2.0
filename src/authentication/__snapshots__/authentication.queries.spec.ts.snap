// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticationQueries composeUserSession should return no user if not found 1`] = `null`;

exports[`AuthenticationQueries composeUserSession should return the correct user 1`] = `
Object {
  "email": "development@icapps.com",
  "firstName": null,
  "lastName": null,
  "permissions": Object {
    "roles": Object {
      "edit": true,
      "view": true,
    },
    "users": Object {
      "edit": true,
      "view": true,
    },
  },
  "state": "ACTIVE",
  "userId": "c4cb4582-1e97-4e3e-9d49-c744c8c1a229",
}
`;

exports[`AuthenticationQueries getLoggedInUser should return no user if not found 1`] = `undefined`;

exports[`AuthenticationQueries getLoggedInUser should return the correct user 1`] = `
Object {
  "createdAt": 2020-01-01T08:21:59.730Z,
  "createdBy": null,
  "email": "development@icapps.com",
  "firstName": null,
  "id": "c4cb4582-1e97-4e3e-9d49-c744c8c1a229",
  "lastName": null,
  "permissions": Object {
    "roles": Object {
      "edit": true,
      "view": true,
    },
    "users": Object {
      "edit": true,
      "view": true,
    },
  },
  "state": "ACTIVE",
  "updatedAt": 2020-01-01T08:21:59.730Z,
  "updatedBy": null,
}
`;
